lib/tcpip/tcpip_error_handler.h

```diff
@@ -0,0 +1,366 @@
+extern std::atomic<uint32_t> currentSocketEpoch;
+extern std::atomic<uint8_t>  userThreadCount;
+
+// Locks and futexes to release as part of the reset
+extern struct FlagLockState       ipThreadLockState;
+extern struct RecursiveMutexState __CriticalSectionFlagLock;
+extern struct RecursiveMutexState __SuspendFlagLock;
+extern QueueHandle_t              xNetworkEventQueue;
+
+// APIs to call as part of the reset
+extern void free_buffer_manager_memory();
+extern void network_restart();
+
+/// Thread ID of the network thread.
+extern uint16_t networkThreadID;
+
+/**
+ * Free the compartment's heap memory.
+ *
+ * Note that socket memory will not be freed because sockets are allocated with
+ * user-passed capabilities which we do not store. API users are supposed to
+ * close them, which will trigger a free.
+ */
+__always_inline static inline void free_compartment_memory()
+{
+	// Global heap capability.
+	heap_free_all(MALLOC_CAPABILITY);
+	// Buffer manager capability. If the buffer manager is using the global
+	// heap capability, this will do nothing.
+	free_buffer_manager_memory();
+}
+
+/**
+ * Reset the network stack state.
+ *
+ * We go through all locks used in the TCP/IP compartment and set them for
+ * destruction. The list of synchronization primitives resetted here was
+ * extracted through a manual study of the compartment's code-base: this may
+ * therefore break if new releases of FreeRTOS+TCP introduce new locks. In the
+ * future, we may want to come up with a more systematic approach.
+ *
+ * This function is designed to be robust against most types of compartment
+ * corruption, however we do assume that:
+ * - 'reset-critical' data has not been corrupted
+ * - the control-flow of threads in the compartment has not been altered
+ */
+extern "C" void reset_network_stack_state()
+{
+	const bool isUserThread = thread_id_get() != networkThreadID;
+	const bool isIpThread   = !isUserThread;
+
+	if (isUserThread)
+	{
+		DebugErrorHandler::log(
+		  "User thread TCP/IP stack error handler called!");
+		userThreadCount--;
+	}
+	else
+	{
+		DebugErrorHandler::log(
+		  "Network thread TCP/IP stack error handler called!");
+	}
+
+	// Manually unlock the sealed sockets lock list if it was held.
+	if (sealedSocketsListLock.get_owner_thread_id() == thread_id_get())
+	{
+		DebugErrorHandler::log("The sealed sockets lock was held by the "
+		                       "crashing thread. Forcefully unlocking it.");
+		sealedSocketsListLock.unlock();
+	}
+
+	// Set the currently restarting flag. This will do several things:
+	// 1. ensure that only one call to this error handler triggers a reset
+	// 2. ensure that no thread enters the compartment while we are
+	//    restarting
+	// 3. reset the network thread whenever it wakes up
+	uint32_t expected = NotRestarting;
+	if (!restartState.compare_exchange_strong(expected, Restarting))
+	{
+		if (isIpThread && ((restartState.load() & IpThreadKicked) != 0))
```

Comment:
I think it's possible that `restartState` has been altered between the `compare_exchange_strong` and here, so `IpThreadKicked` may have been asserted and isn't now.  That's probably fine, but maybe this could be

```suggestion
		if (isIpThread && ((expected & IpThreadKicked) != 0))
```

Reply:
Actually I think that if the flag was cleared, it's because the concurrent restart actually finished between the two loads. In that case we do not want to print the log message. Still, I think that this is an extremely unlikely thing to happen (if it can happen at all), and it's a good idea to get rid of that load as you suggest.